### ISSUE TITLE
Unable to run on qBittorrent 4.5.0

### ISSUE THREAD

**Issue by PozoSer** [jerrymakesjelly/autoremove-torrents#157]
The script properly identifies the torrents to be deleted but seems unable to do so. It returns: The torrent XX and its data cannot be removed. Reason: The server responses HTTP 405. A quick search on Google indicates that a change introduced in version 4.4.5 now require "All API methods only allows GET or POST methods. Use POST when you are mutating some state (or when your request is too big to fit into GET) and use GET otherwise. Starting with qBittorrent v4.4.4, server will return 405 Method Not Allowed when you used the wrong request method."

https://github.com/qbittorrent/qBittorrent/wiki/WebUI-API-(qBittorrent-4.1)

**Comment by Sleashe** [comment]
Yup, getting same here !

**Comment by Jarsky** [comment]
Just confirming also have the same error with qBitTorrent 4.5.0. Upgraded from 4.3.9 which was working fine. 

`Tue, 20 Dec 2022 08:35:02 autoremovetorrents.task ERROR: The torrent xxxxxx and its data cannot be removed. Reason: The server responses HTTP 405`

**Comment by PozoSer** [comment]
I'm still getting the same issue with the @Jorman patch.

**Comment by Jarsky** [comment]
> I'm still getting the same issue with the @Jorman patch.

Did you use the patch-1 branch? 
I can confirm that it is working with Jorman's patch

![image](https://user-images.githubusercontent.com/839416/209305024-de0374e7-1a00-40fe-b813-294faa34c01e.png)

**Comment by PozoSer** [comment]
Yeah. Tried deleting everything and re-downloading but still getting the same error.

![imagen](https://user-images.githubusercontent.com/18380820/210365086-91ac0fd8-e6eb-4b59-b3e7-589862a8a789.png)

I have manually checked that the changes Jarsky did to the qBittorrent client in lines 113 and 119 are in fact downloaded and it seems like it. These are the only changes in the patch as far as I've noticed.

![imagen](https://user-images.githubusercontent.com/18380820/210365288-5e786e5c-6a7c-41ad-9269-4ca5408fced5.png)

**Comment by Jorman** [comment]
@PozoSer 
Yes, is the only modification.
You can also modify it by hand, for example I use autoremove-torrents with python venv, so I manually edited the file, but I remember I also update python version inside venv, so in this case I've two dirctory for autoremove-torrent, the old one and the new one.
How do you have configured it?
qBittorrent version?
Phyton version?

**Comment by PozoSer** [comment]
Thanks Jorman for the interest and help. I finally got it to work by manually uninstalling it and downloading it again. For some reason overwriting the autoremove-torrents folder was not working.

**Comment by Alcedema** [comment]
I'm still receiving the 405 error, even when using the new patch-1 branch.

I uninstalled the old one completely; 

```
python3 setup.py install --record files.txt
cat files.txt | xargs rm -rf
rm autoremove-torrents
which autoremove-torrents

```

Then installed the Jorman patch (and confirmed modifications);

```
git clone -b patch-1 http://github.com/Jorman/autoremove-torrents.git
cd autoremove-torrents
awk 'NR==113' ./autoremovetorrents/client/qbittorrent.py 
            _return self._session.post(self._host+'/api/v2/torrents/delete', data={'hashes':'|'.join(torrent_hash_list), 'deleteFiles': False})_
python3 setup.py install
which autoremove-torrents 
            _/usr/bin/autoremove-torrents_
```

Ran ART (after --view to confirm);

`/usr/bin/autoremove-torrents --conf=/etc/autotorrent/config.yml`

```
Sat, 04 Mar 2023 09:29:05 autoremovetorrents.main INFO: Auto Remove Torrents 1.5.4
Sat, 04 Mar 2023 09:29:05 autoremovetorrents.main INFO: Loading configurations...
Sat, 04 Mar 2023 09:29:05 autoremovetorrents.main INFO: Found 1 task(s) in the file.
Sat, 04 Mar 2023 09:29:05 autoremovetorrents.task INFO: Running task 'my_task'...
Sat, 04 Mar 2023 09:29:05 autoremovetorrents.task INFO: Logging in...
Sat, 04 Mar 2023 09:29:05 autoremovetorrents.task INFO: Login successfully. The client is qBittorrent v4.5.2.
Sat, 04 Mar 2023 09:29:05 autoremovetorrents.task INFO: WebUI API version: 2.8.19 (v2)
Sat, 04 Mar 2023 09:29:08 autoremovetorrents.task INFO: Status reported by the client: 
...
Sat, 04 Mar 2023 09:30:02 autoremovetorrents.strategy INFO: Total: 1 torrent(s). 1 torrent(s) can be removed.
Sat, 04 Mar 2023 09:30:02 autoremovetorrents.strategy INFO: To be deleted:
...
Sat, 04 Mar 2023 09:30:03 autoremovetorrents.task ERROR: The torrent REDACTED and its data cannot be removed. Reason: The server responses HTTP 405.
```

**Comment by Jorman** [comment]
Hi, you've to clone the patch version in order to use it, see my log:
```
Thu, 28 Sep 2023 14:25:51 main.py[line:51] INFO Auto Remove Torrents 1.5.4
Thu, 28 Sep 2023 14:25:52 main.py[line:53] INFO Loading configurations...
Thu, 28 Sep 2023 14:25:52 main.py[line:56] INFO Found 1 task(s) in the file.
Thu, 28 Sep 2023 14:25:52 task.py[line:138] INFO Running task 'qBittorrent'...
Thu, 28 Sep 2023 14:25:52 task.py[line:85] INFO Logging in...
Thu, 28 Sep 2023 14:25:52 task.py[line:87] INFO Login successfully. The client is qBittorrent v4.5.5.
Thu, 28 Sep 2023 14:25:52 task.py[line:88] INFO WebUI API version: 2.8.19 (v2)
Thu, 28 Sep 2023 14:25:52 task.py[line:92] INFO Status reported by the client: 
	Download Speed: 0.00B/s	Total: 73.17GiB
	Upload Speed: 0.00B/s	Total: 161.99GiB
Thu, 28 Sep 2023 14:25:52 task.py[line:96] INFO Getting all the torrents...
Thu, 28 Sep 2023 14:25:53 task.py[line:103] INFO Please wait...We have found 82 torrent(s).
Thu, 28 Sep 2023 14:25:53 task.py[line:106] INFO Found 142 torrent(s) in the client.
Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker1...
Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker2...
Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker3...
Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker4...
Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker5...
Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker6...
Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker7...
Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker8...
Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Private-Tracker9...
Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 3 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Private-10...
Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Private-Tracker11...
Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 23 torrent(s). 1 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:54 strategy.py[line:177] INFO To be deleted:
Thu, 28 Sep 2023 14:25:54 strategy.py[line:179] INFO xxx
	Progress:100.00%	Size:19.46GiB	Ratio:0.354	Total Uploaded:6.88GiB
	Seeder(connected/total):0/31	Leecher(connected/total):1/1	Status:Uploading
	Download Speed:0.00B/s(Avg.:34.54MiB/s)	Upload Speed:0.00B/s(Avg.:9.14KiB/s)
	Create Time:2023-09-19 11:13:05	Seeding Time:9d 03:02:55	Downloading Time:(Not Provided)	Last Activity:1d 00:41:39
	Category:FreeLeech	Tracker:** [DHT] **,** [PeX] **,** [LSD] **,xxx
Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Private-Tracker12...
Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Global-Tracker...
Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
Thu, 28 Sep 2023 14:25:54 task.py[line:125] INFO The torrent xxx and its data have been removed.
```

**Comment by speedster-pta** [comment]
> Hi, you've to clone the patch version in order to use it, see my log:
> 
> ```
> Thu, 28 Sep 2023 14:25:51 main.py[line:51] INFO Auto Remove Torrents 1.5.4
> Thu, 28 Sep 2023 14:25:52 main.py[line:53] INFO Loading configurations...
> Thu, 28 Sep 2023 14:25:52 main.py[line:56] INFO Found 1 task(s) in the file.
> Thu, 28 Sep 2023 14:25:52 task.py[line:138] INFO Running task 'qBittorrent'...
> Thu, 28 Sep 2023 14:25:52 task.py[line:85] INFO Logging in...
> Thu, 28 Sep 2023 14:25:52 task.py[line:87] INFO Login successfully. The client is qBittorrent v4.5.5.
> Thu, 28 Sep 2023 14:25:52 task.py[line:88] INFO WebUI API version: 2.8.19 (v2)
> Thu, 28 Sep 2023 14:25:52 task.py[line:92] INFO Status reported by the client: 
> 	Download Speed: 0.00B/s	Total: 73.17GiB
> 	Upload Speed: 0.00B/s	Total: 161.99GiB
> Thu, 28 Sep 2023 14:25:52 task.py[line:96] INFO Getting all the torrents...
> Thu, 28 Sep 2023 14:25:53 task.py[line:103] INFO Please wait...We have found 82 torrent(s).
> Thu, 28 Sep 2023 14:25:53 task.py[line:106] INFO Found 142 torrent(s) in the client.
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker1...
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker2...
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker3...
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker4...
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker5...
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker6...
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker7...
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:53 strategy.py[line:167] INFO Running strategy Private-Tracker8...
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Private-Tracker9...
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 3 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Private-10...
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Private-Tracker11...
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 23 torrent(s). 1 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:177] INFO To be deleted:
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:179] INFO xxx
> 	Progress:100.00%	Size:19.46GiB	Ratio:0.354	Total Uploaded:6.88GiB
> 	Seeder(connected/total):0/31	Leecher(connected/total):1/1	Status:Uploading
> 	Download Speed:0.00B/s(Avg.:34.54MiB/s)	Upload Speed:0.00B/s(Avg.:9.14KiB/s)
> 	Create Time:2023-09-19 11:13:05	Seeding Time:9d 03:02:55	Downloading Time:(Not Provided)	Last Activity:1d 00:41:39
> 	Category:FreeLeech	Tracker:** [DHT] **,** [PeX] **,** [LSD] **,xxx
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Private-Tracker12...
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:167] INFO Running strategy Global-Tracker...
> Thu, 28 Sep 2023 14:25:54 strategy.py[line:174] INFO Total: 0 torrent(s). 0 torrent(s) can be removed.
> Thu, 28 Sep 2023 14:25:54 task.py[line:125] INFO The torrent xxx and its data have been removed.
> ```

Thanks for the feedback. Are you perhaps able to help me (or point me to a link) about how the cloning works?

**Comment by Jorman** [comment]
I personally use venv, I prefer to "trap" each python project, so maybe is not so easy
You can also modify the file by hand, but you've to modify the right one.
I use python 3.9 inside venv, pay attention because is possible that you've more than one version of python and you've to modify the file for the version you use, for this I only use venv :)

If you want I can explain how I personally use venv for this project

**Comment by Jorman** [comment]
The simple way, for my point of view, is to navigate where do you want to store the venv
1. Prepare the venv
`python3 -m venv autoremove-torrent`
2. Activate the Virtual Environment
`source autoremove-torrent/bin/activate`
3. Install the git version with patch
`pip install git+https://github.com/Jorman/autoremove-torrents@patch-1`
4. Install/upgrade pip wheel and setuptools (maybe not necessary but I prefer to)
`pip install --upgrade pip setuptools wheel`
5. Deactivate the Virtual Environment (because I use it not from inside the Virtual Environment but from outsie)
`deactivate`
6. Now you can test the program with
`/path_to_the_venv_autoremove_torrent_folder/bin/autoremove-torrents --conf=/path_to_your_config_file/config.yml --view --debug`
7. To use the program from outside
`/path_to_the_venv_autoremove_torrent_folder/autoremove-torrents --conf=/path_to_your_config_file/config.yml --log=/path_where_you_want_to_save_the_logs/logs`

**Comment by blastbeng** [comment]
I am still experiencing this issue with qbit v4.6.2.

I have installed the repo version with patch:
`pip install git+https://github.com/Jorman/autoremove-torrents@patch-1`

Does anyone ever solved this problem?

**Comment by Jorman** [comment]
@blastbeng 
Do you removed all the old instance?
Do you have more than one python version installed?
Follow my instruction to use the program inside a venv environment, otherwise you've to be sure to install and use the right version, for example for python 3 etc.
I use the patched version for the latest qBittorrent and is still working, no problem at all

**Comment by blastbeng** [comment]
@Jorman 
I have solved using this project inside a python project:

https://github.com/blastbeng/qbittorrent-fix-stalled

I import your lib and i manually launch the  process by a scheduler or from a rest API..

**Comment by Jorman** [comment]
@blastbeng 
Glad to know that you resolved, I don't use python project, I use the venv functionality to use all python script.
Please take note that I only made a patch, this project remains the same as before so logs are in the same place, like if you use the main jerrymakesjelly project. I'm not an expert on python projects but I think it is normal that all remain inside the project. Maybe check the wiki from the main project to know how to manage logs, I personally use a systemd timer with a script that check hardlinks and launch specific autoremove-torrent configuration and all the python project are inside the venv environment

**Comment by Morcegolas** [comment]
Hello.
First let me thanks Jorman for resolving this problem.
Second I must ask for anybody that can help me, I'm using docker container to autoremove-torrents on my NAS, but I don't know how to update qbittorrent.py. I already tried to run 'find . -name qbittorrent.py' inside the container but without any result.
If anybody can help me I'll be grateful, I already searched but can't find any updated container that have this patch.
